kyuubi-common/src/main/scala/org/apache/kyuubi/service/TBinaryFrontendService.scala

```diff
@@ -0,0 +1,116 @@
+  extends TFrontendService(name) with TCLIService.Iface with Runnable with Logging {
+
+  import KyuubiConf._
+
+  /**
+   * @note this is final because we don't want new implementations for engine to override this.
+   *       and we shall simply set it to zero for randomly picking an available port
+   */
+  final override protected val portKey: ConfigEntry[Int] = FRONTEND_THRIFT_BINARY_BIND_PORT
+
+  private var server: Option[TServer] = None
+
+  // When a OOM occurs, here we de-register the engine by stop its discoveryService.
+  // Then the current engine will not be connected by new client anymore but keep the existing ones
+  // alive. In this case we can reduce the engine's overhead and make it possible recover from that.
+  // We shall not tear down the whole engine by serverable.stop to make the engine unreachable for
+  // the existing clients which are still getting statuses and reporting to the end-users.
+  protected def oomHook: Runnable = {
+    () => discoveryService.foreach(_.stop())
+  }
+
+  override def initialize(conf: KyuubiConf): Unit = synchronized {
+    this.conf = conf
+    try {
+      val minThreads = conf.get(FRONTEND_THRIFT_MIN_WORKER_THREADS)
+      val maxThreads = conf.get(FRONTEND_THRIFT_MAX_WORKER_THREADS)
+      val keepAliveTime = conf.get(FRONTEND_THRIFT_WORKER_KEEPALIVE_TIME)
+      val executor = ExecutorPoolCaptureOom(
+        name + "Handler-Pool",
+        minThreads,
+        maxThreads,
+        keepAliveTime,
+        oomHook)
+      val transFactory = authFactory.getTTransportFactory
+      val tProcFactory = authFactory.getTProcessorFactory(this)
+      val tServerSocket = new TServerSocket(serverSocket)
+      val maxMessageSize = conf.get(FRONTEND_THRIFT_MAX_MESSAGE_SIZE)
+      val requestTimeout = conf.get(FRONTEND_THRIFT_LOGIN_TIMEOUT).toInt
+      val beBackoffSlotLength = conf.get(FRONTEND_THRIFT_LOGIN_BACKOFF_SLOT_LENGTH).toInt
+      val args = new TThreadPoolServer.Args(tServerSocket)
+        .processorFactory(tProcFactory)
+        .transportFactory(transFactory)
+        .protocolFactory(new TBinaryProtocol.Factory)
+        .inputProtocolFactory(
+          new TBinaryProtocol.Factory(true, true, maxMessageSize, maxMessageSize))
+        .requestTimeout(requestTimeout).requestTimeoutUnit(TimeUnit.MILLISECONDS)
+        .beBackoffSlotLength(beBackoffSlotLength)
+        .beBackoffSlotLengthUnit(TimeUnit.MILLISECONDS)
+        .executorService(executor)
+      // TCP Server
+      server = Some(new TThreadPoolServer(args))
+      server.foreach(_.setServerEventHandler(new FeTServerEventHandler))
+      info(s"Initializing $name on ${serverAddr.getHostName}:${serverSocket.getLocalPort} with" +
+        s" [$minThreads, $maxThreads] worker threads")
+    } catch {
+      case e: Throwable =>
+        error(e)
+        throw new KyuubiException(
+          s"Failed to initialize frontend service on $serverAddr:$portNum.",
+          e)
+    }
+    super.initialize(conf)
```

Comment:
move to line 91 ?

Reply:
playing trick here as we always call `super.initialize` at the beginning or at the end